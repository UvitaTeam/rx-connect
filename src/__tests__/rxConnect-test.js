import React from "react";
import Rx from "rx";
import renderer from "react-test-renderer";

import { rxConnect } from "../";

test("works with Observable", () => {
    const props$ = Rx.Observable.of({ a: 123 });

    const Component = rxConnect(props$)(({ a }) => <div>{a}</div>);

    const tree = renderer.create(<Component />).toJSON();

    expect(tree).toMatchSnapshot();
});

test("works with Array", () => {
    const props$ = () => [
        Rx.Observable.of({ a: 123 }),
        Rx.Observable.of({ foo: "bar" }),
    ];

    const Component = rxConnect(props$)(({ a, foo }) => <div>{a}{foo}</div>);

    const tree = renderer.create(<Component />).toJSON();

    expect(tree).toMatchSnapshot();
});

test("works with Generator", () => {
    const props$ = function* () {
        yield Rx.Observable.of({ a: 123 });
        yield Rx.Observable.of({ foo: "bar" });
    };

    const Component = rxConnect(props$)(({ a, foo }) => <div>{a}{foo}</div>);

    const tree = renderer.create(<Component />).toJSON();

    expect(tree).toMatchSnapshot();
});

test("passes properties as Observable", () => {
    const connector = props$ => props$.pluck("someProp").map(a => ({ a }));

    const Component = rxConnect(connector)(({ a }) => <div>{a}</div>);

    const tree = renderer.create(<Component someProp={123} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test("passes children automatically", () => {
    const Component = rxConnect(Rx.Observable.of({}))(({ children }) => <div>{children}</div>);

    const tree = renderer.create(<Component>Hello, RxConnect!</Component>).toJSON();

    expect(tree).toMatchSnapshot();
});

test("ignores not connect properties", () => {
    const Component = rxConnect(Rx.Observable.of({ }))(({ a }) => <div>{a}</div>);

    const tree = renderer.create(<Component a={123} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test("accepts function-based mutations", () => {
    const Component = rxConnect(Rx.Observable.of(() => ({ a: 123 })))(({ a }) => <div>{a}</div>);

    const tree = renderer.create(<Component />).toJSON();

    expect(tree).toMatchSnapshot();
});

test("throws an error if mutation is neither an object or function", () => {
    // eslint-disable-next-line no-console
    console.error = jest.genMockFn();

    const Component = rxConnect(Rx.Observable.of([ 123 ]))(({ a }) => <div>{a}</div>);
    renderer.create(<Component />).toJSON();

    // eslint-disable-next-line no-console
    expect(console.error.mock.calls[0]).toMatchSnapshot();
});

test("receives new props", async () => {
    const connector = props$ => props$.map(({ a, b }) => ({ a: a + b }));

    const Component = rxConnect(connector)(({ a }) => <div>{a}</div>);

    class Parent extends React.Component {
        state = {
            a: 10,
            b: 5
        };

        render() {
            return <Component {...this.state} />
        }
    }

    const parent = renderer.create(<Parent />);

    // 15
    expect(parent.toJSON()).toMatchSnapshot();

    parent.getInstance().setState({ a: -5 });

    // Still 15 because of debouncing
    expect(parent.toJSON()).toMatchSnapshot();

    parent.getInstance().setState({ b: -10 });

    // Skip 1 frame because of debounce
    await Rx.Observable.interval(0).take(1).toPromise();

    expect(parent.toJSON()).toMatchSnapshot();
});

test("noDebounce", () => {
    const Component = rxConnect(props$ => props$, { noDebounce: true })(({ i }) => <div>{i}</div>);

    class Parent extends React.Component {
        state = {
            i: 0
        };

        render() {
            return <Component {...this.state} />
        }
    }

    const parent = renderer.create(<Parent />);

    expect(parent.toJSON()).toMatchSnapshot();

    parent.getInstance().setState({ i: 1 });

    expect(parent.toJSON()).toMatchSnapshot();
})

test("handles unmount", () => {
    const props$ = new Rx.ReplaySubject();

    expect(props$.hasObservers()).toBeFalsy();

    const Component = rxConnect(props$)(({ a }) => <div>{a}</div>);

    const node = renderer.create(<Component />);

    expect(props$.hasObservers()).toBeTruthy();

    node.unmount();

    expect(props$.hasObservers()).toBeFalsy();
});
