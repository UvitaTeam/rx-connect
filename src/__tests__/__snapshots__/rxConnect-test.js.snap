exports[`test accepts function-based mutations 1`] = `
<div>
  123
</div>
`;

exports[`test ignores not connect properties 1`] = `<div />`;

exports[`test noDebounce 1`] = `
<div>
  0
</div>
`;

exports[`test noDebounce 2`] = `
<div>
  1
</div>
`;

exports[`test passes children automatically 1`] = `
<div>
  Hello, RxConnect!
</div>
`;

exports[`test passes properties as Observable 1`] = `
<div>
  123
</div>
`;

exports[`test receives new props 1`] = `
<div>
  15
</div>
`;

exports[`test receives new props 2`] = `
<div>
  15
</div>
`;

exports[`test receives new props 3`] = `
<div>
  -15
</div>
`;

exports[`test throws an error if mutation is neither an object or function 1`] = `
Array [
  "Mutation must be a plain object or function. Check rxConnect of Component. Got: ",
  Array [
    123,
  ],
]
`;

exports[`test works with Array 1`] = `
<div>
  123
  bar
</div>
`;

exports[`test works with Generator 1`] = `
<div>
  123
  bar
</div>
`;

exports[`test works with Observable 1`] = `
<div>
  123
</div>
`;
